linkerd/duplex/src/lib.rs

```diff
@@ -85,81 +97,170 @@ where
             is_shutdown: false,
             io,
             direction,
+            flushing: false,
         }
     }
 
-    fn copy_into<U>(
+    /// Reads data from `self`, buffering it, and writing it to `dst.
+    ///
+    /// Returns ready when the stream has shutdown such that no more data may be
+    /// proxied.
+    fn copy_into<U: AsyncWrite + Unpin>(
         &mut self,
         dst: &mut HalfDuplex<U>,
         cx: &mut Context<'_>,
-    ) -> Poll<Result<(), io::Error>>
-    where
-        U: AsyncWrite + Unpin,
-    {
+    ) -> io::Poll<()> {
         // Since Duplex::poll() intentionally ignores the Async part of our
         // return value, we may be polled again after returning Ready, if the
         // other half isn't ready. In that case, if the destination has
         // shutdown, we finished in a previous poll, so don't even enter into
         // the copy loop.
         if dst.is_shutdown {
+            // A prior shutdown may not have completed, so continue driving the
+            // shutdown.
             trace!(direction = %self.direction, "already shutdown");
-            return Poll::Ready(Ok(()));
+            return Pin::new(&mut dst.io).poll_shutdown(cx);
```

Comment:
Actually, `dst.is_shutdown` can't be true unless shutdown completed. this should go back to how it was.